File: 2021/09/02/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        a = {}
        def dfs(low, high):
            if low > high:
                return [None]
            if (low, high) in a:
                return a[(low,high)]
            ans = []
            for mid in range(low, high + 1):
                left = dfs(low, mid - 1)
                right = dfs(mid + 1, high)
                for l in left:
                    for r in right:
                        root = TreeNode(mid)
                        root.left = l
                        root.right = r
                        ans.append(root)
            a[(low,high)]=ans
            return ans
        return dfs(1, n)

File: 2021/09/02/main.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func generateTrees(n int) []*TreeNode {
	var dfs func(int, int) []*TreeNode

	dfs = func(low, high int) []*TreeNode {
		if low > high {
			return []*TreeNode{nil}
		}
		ans := []*TreeNode{}
		for mid := low; mid <= high; mid++ {
			left := dfs(low, mid-1)
			right := dfs(mid+1, high)
			for _, l := range left {
				for _, r := range right {
					root := &TreeNode{mid, nil, nil}
					root.Left = l
					root.Right = r
					ans = append(ans, root)
				}
			}
		}
		return ans
	}
	return dfs(1, n)
}
